homeassistant/components/sensorpush_cloud/api.py

```diff
@@ -0,0 +1,155 @@
+"""Library for interfacing with the SensorPush Cloud API."""
+
+from __future__ import annotations
+
+from collections.abc import Callable
+from datetime import datetime
+from functools import partial, wraps
+import json
+from threading import Lock
+from typing import Any, Concatenate
+
+from sensorpush_api import (
+    AccessTokenRequest,
+    ApiApi,
+    ApiClient,
+    AuthorizeRequest,
+    Configuration,
+    Samples,
+    SamplesRequest,
+    Sensors,
+    SensorsRequest,
+)
+from sensorpush_api.rest import ApiException
+
+from homeassistant.core import HomeAssistant
+from homeassistant.util import dt as dt_util
+
+from .const import ACCESS_TOKEN_EXPIRATION, LOGGER, REQUEST_RETRIES, REQUEST_TIMEOUT
+
+
+def api_call[**_P, _R](
+    func: Callable[Concatenate[SensorPushCloudApi, _P], _R],
+) -> Callable[Concatenate[SensorPushCloudApi, _P], _R]:
+    """Decorate API calls to handle SensorPush Cloud exceptions."""
+
+    @wraps(func)
+    def _api_call(self: SensorPushCloudApi, *args: _P.args, **kwargs: _P.kwargs) -> _R:
+        LOGGER.debug(f"API call to {func} with args={args}, kwargs={kwargs}")
+        retries: int = 0
+        while True:
+            try:
+                result = func(self, *args, **kwargs)
+            except Exception as e:
```

Comment:
Catching the rare `Exception` outside the config flow is not allowed. Please adopt the code to catch only specific ones